p2p/socket.go

```diff
@@ -0,0 +1,461 @@
+
+	buff := s.c.encode(Binary, iserroreof, reqnum, b, wrapped)
+	(*writeBuffer) = append((*writeBuffer), buff...)
+
+	s.buffers.write.Signal()
+	return uint(len(b)), nil // TODO(derrandz): should length be of b or of the encoded b
+}
+
+// writeChunkAckful is a writeChunk that expects to receive an ACK response for the chunk it has written
+// This method will create a request, which is basically a nonce to identify the chunk to write, and a channel on which to receive the response
+// the channel is blocking, thus allowing the 'wait to receive the response' behavior.
+// the `read` routine takes care of identifying incoming responses (_using the nonce_) and redirecting them to the waiting channels of the currently-open requests.
+func (s *socket) writeChunkAckful(b []byte, wrapped bool) (types.Work, error) {
+	request := s.requests.Get()
+	requestNonce := request.Nonce()
+
+	if _, err := s.writeChunk(b, false, requestNonce, wrapped); err != nil {
+		s.requests.Delete(requestNonce)
+		return types.NewWork(requestNonce, nil, "", false), err
+	}
+
+	var response types.Work
+
+	select {
+	case response = <-request.Response():
+		return response, nil
+
+	case <-time.After(time.Millisecond * time.Duration(s.readTimeout)):
+		return types.Work{}, ErrSocketRequestTimedOut(s.addr, requestNonce)
+	}
+}
+
+// the write routine
+// this routine performs buffered writes (writeBufferSize) on the established connection
+// this routine halfs if:
+//   - the cancelable routine cancels
+//   - if the runner stops
+//   - if the socket closes
+//   - if some party signals that this routine should stop (by closing the s.writing channel)
+//   - if there is a write error
+func (s *socket) write(ctx context.Context) {
+	defer func() {
+		close(s.writing)
+		s.closed <- 1
+	}()
+
+	{
+		s.writing <- 1 // signal start
+	}
+
+writer:
+	for {
+		select {
+		case <-ctx.Done():
+			break writer
+		case <-s.runner.Done():
+			break writer
+
+		case <-s.done:
+			break writer
+
+		case _, open := <-s.writing:
+			if !open {
+				break writer
+			}
+
+		case <-s.buffers.write.Signals(): // blocks
+			{
+				if !s.buffers.write.IsOpen() {
+					break writer
+				}
+
+				buff := s.buffers.write.DumpBytes()
+
+				if _, err := s.writer.Write(buff); err != nil {
+					s.error(err)
+					break writer
+				}
+
+				if err := s.writer.Flush(); err != nil {
+					s.error(err)
+					break writer
+				}
+			}
+		}
+	}
+}
+
+// tracks and stores the encountered error
+func (s *socket) error(err error) {
+	defer s.err.Unlock()
+	s.err.Lock()
+
+	if s.err.error == nil {
+		s.err.error = err
+	}
+
+	s.errored <- 1
+}
+
+// A constructor to create a socket
+func NewSocket(readBufferSize uint, packetHeaderLength uint, readTimeoutInMs uint) *socket {
+	wc := newWireCodec()
+	pipe := &socket{
+		c: wc,
+
+		kind:         UndefinedSocketKind,
+		headerLength: packetHeaderLength,
+		bufferSize:   readBufferSize,
+		readTimeout:  readTimeoutInMs,
+
+		buffers: struct {
```

Comment:
I had a question earlier about why this is a custom inline struct that I want to reiterate here.

However, if it is really needed, we should define a type for it so we don't have to redefine it inline twice (here and in `type socket struct`).

Reply:
Not much reuse anywhere else for it become a separate type, this is a nesting in struct to further organize the struct fields, thus why I kept it inline. If it gets defined one more time elsewhere I will type it away.

Reply:
The same one is used on line 432 and 45.

I think this should be externalized into a local struct.